### ISSUE TITLE
Issue with "data = dde.data.PDE" running speed  and network architecture

### ISSUE THREAD

**Issue by MagedBadawi** [lululxvi/deepxde#1957]
Hi @lululxvi , @praksharmam,
I wish you are doing well, 

Im trying to solve a problem of potential field propagation where i have two different boundary conditions (anode and cathode points which are dirichlet type) and (the rest of the points which are flux/Neumann type). 

After debugging my code, i found that the there is a bottleneck issue with num_domain=len(domain_points), num_boundary=len(boundary_points),  
        anchors=self.coords_po_po. As my understanding they must be set like that but when i try to set num_domain and num_boundary = 2 and  anchors=None as a trial it works fast. 

Could you give me insights about the main issue in this code? 
Also shall i define a third boundary condition (Neumann) for the rest of the points that are main thing to be trained here and if so, how should i do that and what will be the difference between in and Laplace equation used?  

My Code:

class DGMModel:
    def __init__(self, coords, conductivity, potential):

        self.coords = coords
        self.conductivity = conductivity
        self.potential = potential.astype(np.float32)

        #masks to separate anode, cathode, and internal voxels
        anode_mask = (self.potential == 1.0).flatten()  # Select all voxels where potential = 1
        cathode_mask = (self.potential == 0.0).flatten()  # Select all voxels where potential = 0
        internal_mask = (~anode_mask) & (~cathode_mask)  # Select all other voxels (0 < potential < 1)

        # Extraction of data
        self.coords_anode_bc = self.coords[anode_mask]  #Potential = 1
        self.coords_cathode_bc = self.coords[cathode_mask]  #Potential = 0
        self.coords_po_po = self.coords[internal_mask]  #Internal voxels

        self.conductivity_anode_bc = self.conductivity[anode_mask]  #anode conductivity
        self.conductivity_cathode_bc = self.conductivity[cathode_mask]  #cathode conductivity
        self.conductivity_po_po = self.conductivity[internal_mask]  #internal voxels conductivity 

        self.potential_anode_bc = self.potential[anode_mask]  #anode potential
        self.potential_cathode_bc = self.potential[cathode_mask]  #cathode potential 

        print(f" Anode Voxels: {len(self.coords_anode_bc)} | Cathode Voxels: {len(self.coords_cathode_bc)} | Internal Voxels: {len(self.coords_po_po)}", flush=True)

        # dictionaries for lookup
        self.anode_data = {tuple(self.coords_anode_bc[i]): self.potential_anode_bc[i] for i in range(len(self.coords_anode_bc))}
        self.cathode_data = {tuple(self.coords_cathode_bc[i]): self.potential_cathode_bc[i] for i in range(len(self.coords_cathode_bc))}

        self.anode_coords = list(self.anode_data.keys())
        self.cathode_coords = list(self.cathode_data.keys())

        #dataset for fast lookup
        self.conductivity_dict = {
            tuple(self.coords_po_po[i]): float(self.conductivity_po_po[i])  # Ensure float value
            for i in range(len(self.coords_po_po))
        }

        self.model = self.define_model()

    def laplace_pde(self, x, u):
        """Laplace equation sigma * ∇²V = 0 """
        sigma = self.exact_conductivity(x)  
        du_xx = dde.grad.hessian(u, x, i=0, j=0)
        du_yy = dde.grad.hessian(u, x, i=1, j=1)
        du_zz = dde.grad.hessian(u, x, i=2, j=2)

        return sigma * (du_xx + du_yy + du_zz)

    def exact_conductivity(self, x):
        x_tuples = [tuple(pt.detach().cpu().numpy().astype(np.float32)) for pt in x] 
        conductivity_keys = list(self.conductivity_dict.keys())
        kdtree = KDTree(conductivity_keys)
        dist, idx = kdtree.query(x_tuples, k=1)  # Query all at once
        sigma_values = [self.conductivity_dict[conductivity_keys[i]] for i in idx]  
        return torch.tensor(sigma_values, dtype=torch.float32)  

    def exact_anode_potential(self, X):
        return np.ones((len(X), 1), dtype=np.float32)  #anode potential = 1

    def exact_cathode_potential(self, X):
        return np.zeros((len(X), 1), dtype=np.float32)  #cathode potential = 0 

    def boundary_condition_anode(self, x, on_boundary):
        if tuple(x) in self.anode_data:
            return True  
        return False
    def boundary_condition_cathode(self, x, on_boundary):
        if tuple(x) in self.cathode_data:
            return True  
        return False

    def define_model(self):

        domain_points = self.coords_po_po  
        boundary_points = np.vstack((self.coords_anode_bc, self.coords_cathode_bc))

        geom = dde.geometry.PointCloud(domain_points, boundary_points=boundary_points)
        bc_anode = dde.DirichletBC(geom, self.exact_anode_potential, self.boundary_condition_anode)
        bc_cathode = dde.DirichletBC(geom, self.exact_cathode_potential, self.boundary_condition_cathode)
        #bc_flux = dde.NeumannBC(geom, self.compute_flux, self.flux_boundary_condition)
        print("bc_cathode done")
        data = dde.data.PDE(
        geom, 
        self.laplace_pde, 
        [bc_anode, bc_cathode],#, bc_flux],  
        #num_domain=len(domain_points), #len(domain_points),  # Use all domain points
        #num_boundary=len(boundary_points), #len(boundary_points),  # Keep boundary conditions
        anchors=self.coords_po_po #self.coords_po_po  # Force exact dataset points, prevent new sampling
            )
        print("data done")
        net = dde.maps.FNN(
            [3] + [50] * 4 + [1],  
            CONFIG["activation"],  # sigmoid
            CONFIG["Weight Initialization"]  # He uniform
            )
        
        model = dde.Model(data, net)
        model.compile(CONFIG["optimizer"], lr=CONFIG["learning_rate"],loss="MSE", loss_weights=CONFIG["loss_weights"])  #learning_rate = 0.005, loss_weights = [10,1,10] 
        
        return model

**Comment by LordPontiag** [comment]
> Hi [@lululxvi](https://github.com/lululxvi) , @praksharmam, I wish you are doing well,
> 
> Im trying to solve a problem of potential field propagation where i have two different boundary conditions (First: anode and cathode points which are dirichlet type) and (Second: the rest of the points which are flux/Neumann type).
> 
> After debugging my code, i found that there is a bottleneck with num_domain=len(domain_points) =all points including dirichlet & neumann, num_boundary=len(boundary_points)= dirichleit points only, and anchors=self.coords_po_po(all points including dirichlet and neumann). Knowing that the average points in each sample file before normalization are 2million points and after normalization= 1950548 As my understanding they must be set like that but when i try to set num_domain and num_boundary = 2 and anchors=None as a trial it works fast.
> 
> -Could you give me insights about the main how to define num_domain, num_boundary and anchors (the explanation in the documentation is not clear to me)? -Also shall i define a third boundary condition (Neumann) for the rest of the points which are the main points to predict? If yes, how can i do that what parameters should be there?
> 
> Thank you in advance. def zero_flux_bc(x, u, normal, input_): """ Neumann (Flux) boundary condition: n · σ ∇V = 0 """ sigma = exact_conductivity(x) # Get conductivity at x du_dx = dde.grad.jacobian(u, x, i=0, j=0) du_dy = dde.grad.jacobian(u, x, i=1, j=0) du_dz = dde.grad.jacobian(u, x, i=2, j=0)
> 
> ```
>     # Normal flux = dot product of normal vector with gradient
>     flux = sigma * (du_dx * normal[:, 0] + du_dy * normal[:, 1] + du_dz * normal[:, 2])
>     return flux
> 
>     bc_flux = dde.NeumannBC(geom, zero_flux_bc, lambda x, on_boundary: not (boundary_condition_anode(x, on_boundary) or boundary_condition_cathode(x, on_boundary)))
> ```
> 
> My Code:
> 
> class DGMModel: def **init**(self, coords, conductivity, potential):
> 
> ```
>     self.coords = coords
>     self.conductivity = conductivity
>     self.potential = potential.astype(np.float32)
> 
>     #masks to separate anode, cathode, and internal voxels
>     anode_mask = (self.potential == 1.0).flatten()  # Select all voxels where potential = 1
>     cathode_mask = (self.potential == 0.0).flatten()  # Select all voxels where potential = 0
>     internal_mask = (~anode_mask) & (~cathode_mask)  # Select all other voxels (0 < potential < 1)
> 
>     # Extraction of data
>     self.coords_anode_bc = self.coords[anode_mask]  #Potential = 1
>     self.coords_cathode_bc = self.coords[cathode_mask]  #Potential = 0
>     self.coords_po_po = self.coords[internal_mask]  #Internal voxels
> 
>     self.conductivity_anode_bc = self.conductivity[anode_mask]  #anode conductivity
>     self.conductivity_cathode_bc = self.conductivity[cathode_mask]  #cathode conductivity
>     self.conductivity_po_po = self.conductivity[internal_mask]  #internal voxels conductivity 
> 
>     self.potential_anode_bc = self.potential[anode_mask]  #anode potential
>     self.potential_cathode_bc = self.potential[cathode_mask]  #cathode potential 
> 
> 
>     print(f" Anode Voxels: {len(self.coords_anode_bc)} | Cathode Voxels: {len(self.coords_cathode_bc)} | Internal Voxels: {len(self.coords_po_po)}", flush=True)
> 
>     # dictionaries for lookup
>     self.anode_data = {tuple(self.coords_anode_bc[i]): self.potential_anode_bc[i] for i in range(len(self.coords_anode_bc))}
>     self.cathode_data = {tuple(self.coords_cathode_bc[i]): self.potential_cathode_bc[i] for i in range(len(self.coords_cathode_bc))}
> 
>     self.anode_coords = list(self.anode_data.keys())
>     self.cathode_coords = list(self.cathode_data.keys())
> 
>     #dataset for fast lookup
>     self.conductivity_dict = {
>         tuple(self.coords_po_po[i]): float(self.conductivity_po_po[i])  # Ensure float value
>         for i in range(len(self.coords_po_po))
>     }
> 
>     self.model = self.define_model()
> 
> 
> 
> 
> def laplace_pde(self, x, u):
>     """Laplace equation sigma * ∇²V = 0 """
>     sigma = self.exact_conductivity(x)  
>     du_xx = dde.grad.hessian(u, x, i=0, j=0)
>     du_yy = dde.grad.hessian(u, x, i=1, j=1)
>     du_zz = dde.grad.hessian(u, x, i=2, j=2)
> 
>     return sigma * (du_xx + du_yy + du_zz)
> 
> def exact_conductivity(self, x):
>     x_tuples = [tuple(pt.detach().cpu().numpy().astype(np.float32)) for pt in x] 
>     conductivity_keys = list(self.conductivity_dict.keys())
>     kdtree = KDTree(conductivity_keys)
>     dist, idx = kdtree.query(x_tuples, k=1)  # Query all at once
>     sigma_values = [self.conductivity_dict[conductivity_keys[i]] for i in idx]  
>     return torch.tensor(sigma_values, dtype=torch.float32)  
> 
> def exact_anode_potential(self, X):
>     return np.ones((len(X), 1), dtype=np.float32)  #anode potential = 1
> 
> def exact_cathode_potential(self, X):
>     return np.zeros((len(X), 1), dtype=np.float32)  #cathode potential = 0 
> 
> 
> def boundary_condition_anode(self, x, on_boundary):
>     if tuple(x) in self.anode_data:
>         return True  
>     return False
> def boundary_condition_cathode(self, x, on_boundary):
>     if tuple(x) in self.cathode_data:
>         return True  
>     return False
> 
> 
> def define_model(self):
> 
>     domain_points = self.coords_po_po  
>     boundary_points = np.vstack((self.coords_anode_bc, self.coords_cathode_bc))
> 
>     geom = dde.geometry.PointCloud(domain_points, boundary_points=boundary_points)
>     bc_anode = dde.DirichletBC(geom, self.exact_anode_potential, self.boundary_condition_anode)
>     bc_cathode = dde.DirichletBC(geom, self.exact_cathode_potential, self.boundary_condition_cathode)
>     
>     data = dde.data.PDE(
>     geom, 
>     self.laplace_pde, 
>     [bc_anode, bc_cathode],#, bc_flux],  
>     #num_domain=len(domain_points), #len(domain_points),  # Use all domain points
>     #num_boundary=len(boundary_points), #len(boundary_points),  # Keep boundary conditions
>     anchors=self.coords_po_po #self.coords_po_po  # Force exact dataset points, prevent new sampling
>         )
>     net = dde.maps.FNN(
>         [3] + [50] * 4 + [1],  
>         CONFIG["activation"],  # sigmoid
>         CONFIG["Weight Initialization"]  # He uniform
>         )
>     
>     model = dde.Model(data, net)
>     model.compile(CONFIG["optimizer"], lr=CONFIG["learning_rate"],loss="MSE", loss_weights=CONFIG["loss_weights"])  #learning_rate = 0.005, loss_weights = [10,1,1] 
>     
>     return model
> ```

Hello Dear @MagedBadawi,

For your first question, using num_domain and num_boundary provides you with random training points in your domain and on your boundary respectively. On the other hand using anchors let you to select your desired exact points as training points in the domain and on the boundaries.

If it is important to you to select exact points, make a [x, y] array of your points and set them as anchors like `anchors = CoordinatesArray`. then you can optionally use num_boundary and num_domain also as extra training points.
here is an example

```
data = dde.data.PDE(geom,
                   pde,
                   bc,
                   num_domain = 2000,     # it selects points inside domain
                   num_boundary = 500,    # it selects points on boundary
                   train_distribution = "uniform",
                   anchors = CoordinatesArray,   # it selects points in array of CoordinatesArray in format of [x, y]
                   num_test = 2000
                   )
```
